packages/@aws-cdk/sdk-v2-to-v3-adapter/lib/coerce-api-parameters.ts

```diff
@@ -0,0 +1,54 @@
+import { UINT8ARRAY_PARAMETERS } from './parameter-types';
+
+type ApiParameters = { [param: string]: any };
+
+/**
+ * Given a minimal AWS SDKv3 call definition (service, action, parameters),
+ * coerces nested parameter values into a Uint8Array if that's what the SDKv3 expects.
+ */
+export function coerceApiParametersToUint8Array(service: string, action: string, parameters: ApiParameters = {}): ApiParameters {
+  const pathsToCoerce = UINT8ARRAY_PARAMETERS?.[service.toLowerCase()]?.[action.toLowerCase()] ?? [];
+  for (const path of pathsToCoerce) {
+    coerceToUint8Array(parameters, path.split('.'));
+  }
+  return parameters;
+}
+
+/**
+ * Given an object and a path, traverse to the leaf and coerce the value into a Uint8Array.
+ * This function mutates the original object.
+ *
+ * @example
+ *    const obj = { a: { b: { c: '1' } } }
+ *    coerceToUint8Array(obj, ['a', 'b', 'c'])
+ *    assert(obj == {a: {b: {c: new Uint8Array([49]} } })
+ */
+export function coerceToUint8Array(obj: any, path: string[]): any {
+  const currentKey = path?.[0];
+  if (!currentKey) {
+    return;
+  }
+
+  // recurse into array
+  if (currentKey === '*' && Array.isArray(obj)) {
+    obj.map(e => coerceToUint8Array(e, path.slice(1)));
+    return;
+  }
+
+  // Check if we expect the leaf to be an array
+  // We need to change this before traversing down, otherwise we can't mutate
+  if (path.length === 2 && path[1] === '*' && Array.isArray(obj[currentKey])) {
+    obj[currentKey] = obj[currentKey].map((e: any) => new TextEncoder().encode(e));
```

Comment:
Encoding is done in multiple code paths. It would be nicer if it's in one place at the top of the function, which indicates the end of the recursion:

```ts
export function coerceToUint8Array(obj: any, path: string[]): any {
  if (path.length === 0) {
    return coerceValueToUint8Array(obj);
  }
  
  if (path[0] === '*' && Array.isArray(obj)) {
    return obj.map((e) => coerceToUint8Array(e, path.slice(1));
  }

  if (obj && typeof obj === 'object') {
    if (path[0] in obj) {
      obj[path[0]] = coerceToUint8Array(obj[path[0]], path.slice(1));
    }
    return obj;
  }
  return obj; 
}

function coerceValueToUint8Array(x: unknown): Uint8Array {
  if (x instanceof Uint8Array) { 
    return x;
  }
  // FIXME: need to think about Buffer?
  return new TextEncoder().encode(e);
}
```